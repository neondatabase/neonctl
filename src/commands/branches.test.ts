import { describe } from 'vitest';

import { test } from '../test_utils/fixtures';

describe('branches', () => {
  /* list */

  test('list', async ({ testCliCommand }) => {
    await testCliCommand(['branches', 'list', '--project-id', 'test']);
  });

  /* create */

  test('create by default with r/w endpoint', async ({ testCliCommand }) => {
    await testCliCommand([
      'branches',
      'create',
      '--project-id',
      'test',
      '--name',
      'test_branch',
    ]);
  });

  test('create branch and connect with psql', async ({ testCliCommand }) => {
    await testCliCommand([
      'branches',
      'create',
      '--project-id',
      'test',
      '--name',
      'test_branch',
      '--psql',
    ]);
  });

  test('create branch and connect with psql and psql args', async ({
    testCliCommand,
  }) => {
    await testCliCommand([
      'branches',
      'create',
      '--project-id',
      'test',
      '--name',
      'test_branch',
      '--psql',
      '--',
      '-c',
      'SELECT 1',
    ]);
  });

  test('create with readonly endpoint', async ({ testCliCommand }) => {
    await testCliCommand([
      'branches',
      'create',
      '--project-id',
      'test',
      '--name',
      'test_branch',
      '--type',
      'read_only',
    ]);
  });

  test('create without endpoint', async ({ testCliCommand }) => {
    await testCliCommand([
      'branches',
      'create',
      '--project-id',
      'test',
      '--name',
      'test_branch',
      '--no-compute',
    ]);
  });

  test('create with parent by name', async ({ testCliCommand }) => {
    await testCliCommand([
      'branches',
      'create',
      '--project-id',
      'test',
      '--name',
      'test_branch_with_parent_name',
      '--parent',
      'main',
    ]);
  });

  test('create with parent by lsn', async ({ testCliCommand }) => {
    await testCliCommand([
      'branches',
      'create',
      '--project-id',
      'test',
      '--name',
      'test_branch_with_parent_lsn',
      '--parent',
      '0/123ABC',
    ]);
  });

  test('create with parent by timestamp', async ({ testCliCommand }) => {
    await testCliCommand([
      'branches',
      'create',
      '--project-id',
      'test',
      '--name',
      'test_branch_with_parent_timestamp',
      '--parent',
      '2021-01-01T00:00:00.000Z',
    ]);
  });

  test('create with suspend timeout', async ({ testCliCommand }) => {
    await testCliCommand([
      'branches',
      'create',
      '--project-id',
      'test',
      '--name',
      'test_branch_with_suspend_timeout',
      '--suspend-timeout',
      '60',
    ]);
  });

  test('create with fixed size CU', async ({ testCliCommand }) => {
    await testCliCommand([
      'branches',
      'create',
      '--project-id',
      'test',
      '--name',
      'test_branch_with_fixed_cu',
      '--cu',
      '2',
    ]);
  });

  test('create with autoscaled CU', async ({ testCliCommand }) => {
    await testCliCommand([
      'branches',
      'create',
      '--project-id',
      'test',
      '--name',
      'test_branch_with_autoscaling',
      '--cu',
      '0.5-2',
    ]);
  });

  /* delete */

  test('delete by id', async ({ testCliCommand }) => {
    await testCliCommand([
      'branches',
      'delete',
      'br-sunny-branch-123456',
      '--project-id',
      'test',
    ]);
  });

  /* rename */

  test('rename', async ({ testCliCommand }) => {
    await testCliCommand([
      'branches',
      'rename',
      'test_branch',
      'new_test_branch',
      '--project-id',
      'test',
    ]);
  });

  /* set primary */

  test('set primary by id', async ({ testCliCommand }) => {
    await testCliCommand([
      'branches',
      'set-primary',
      'br-sunny-branch-123456',
      '--project-id',
      'test',
    ]);
  });

  /* set default */

  test('set default by id', async ({ testCliCommand }) => {
    await testCliCommand([
      'branches',
      'set-default',
      'br-sunny-branch-123456',
      '--project-id',
      'test',
    ]);
  });

  /* get */

  test('get by id', async ({ testCliCommand }) => {
    await testCliCommand([
      'branches',
      'get',
      'br-sunny-branch-123456',
      '--project-id',
      'test',
    ]);
  });

  test('get by id', async ({ testCliCommand }) => {
    await testCliCommand([
      'branches',
      'get',
      'br-cloudy-branch-12345678',
      '--project-id',
      'test',
    ]);
  });

  test('get by name', async ({ testCliCommand }) => {
    await testCliCommand([
      'branches',
      'get',
      'test_branch',
      '--project-id',
      'test',
    ]);
  });

  test('get by name with numeric name', async ({ testCliCommand }) => {
    await testCliCommand(['branches', 'get', '123', '--project-id', 'test']);
  });

  /* add compute */

  test('add compute', async ({ testCliCommand }) => {
    await testCliCommand([
      'branches',
      'add-compute',
      'test_branch',
      '--project-id',
      'test',
    ]);
  });

  test('add compute with fixed size CU', async ({ testCliCommand }) => {
    await testCliCommand([
      'branches',
      'add-compute',
      'test_branch_with_fixed_cu',
      '--project-id',
      'test',
      '--cu',
      '2',
    ]);
  });

  test('add compute with autoscaled CU', async ({ testCliCommand }) => {
    await testCliCommand([
      'branches',
      'add-compute',
      'test_branch_with_autoscaling',
      '--project-id',
      'test',
      '--cu',
      '0.5-2',
    ]);
  });

  /* reset */

  test('reset branch to parent', async ({ testCliCommand }) => {
    await testCliCommand([
      'branches',
      'reset',
      'test_branch',
      '--project-id',
      'test',
      '--parent',
    ]);
  });

  /* restore */

  test('restore branch to lsn', async ({ testCliCommand }) => {
    await testCliCommand(
      [
        'branches',
        'restore',
        'br-self-tolsn-123456',
        '^self@0/123ABC',
        '--project-id',
        'test',
        '--preserve-under-name',
        'backup',
      ],
      {
        mockDir: 'restore',
      },
    );
  });

  test('restore to parent branch timestamp by name', async ({
    testCliCommand,
  }) => {
    await testCliCommand(
      [
        'branches',
        'restore',
        'parent-tots',
        '^parent@2021-01-01T00:00:00.000Z',
        '--project-id',
        'test',
      ],
      {
        mockDir: 'restore',
      },
    );
  });

  test('restore to another branch head', async ({ testCliCommand }) => {
    await testCliCommand(
      [
        'branches',
        'restore',
        'br-another-branch-123456',
        'br-any-branch-123456',
        '--project-id',
        'test',
      ],
      {
        mockDir: 'restore',
      },
    );
  });

  test('restore with unexisted branch outputs error', async ({
    testCliCommand,
  }) => {
    await testCliCommand(
      [
        'branches',
        'restore',
        'unexisting-branch',
        '^parent',
        '--project-id',
        'test',
      ],
      {
        mockDir: 'restore',
        code: 1,
        stderr: `ERROR: Branch unexisting-branch not found.
               Available branches: self-tolsn-123456, any-branch, parent-tots, another-branch`,
      },
    );
  });
});
