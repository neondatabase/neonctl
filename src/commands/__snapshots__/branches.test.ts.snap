// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`branches > add compute 1`] = `
"id: test_endpoint_id
branch_id: br-sunny-branch-123456
created_at: 2019-01-01T00:00:00Z
type: read_only
"
`;

exports[`branches > add compute with autoscaled CU 1`] = `
"id: test_endpoint_id
branch_id: br-numbered-branch-123456
created_at: 2019-01-01T00:00:00Z
type: read_only
"
`;

exports[`branches > add compute with fixed size CU 1`] = `
"id: test_endpoint_id
branch_id: br-numbered-branch-123456
created_at: 2019-01-01T00:00:00Z
type: read_only
"
`;

exports[`branches > create branch and connect with psql 1`] = `
"branch:
  id: br-new-branch-123456
  name: test_branch
  created_at: 2021-01-01T00:00:00.000Z
endpoints:
  - id: ep-undefined-123456
    type: read_write
    created_at: 2021-01-01T00:00:00.000Z
    host: undefined.example.com
connection_uris:
  - connection_uri: postgres://ep-undefined-123456.example.com:5432/test_project
{"psql-cli-args":["postgres://ep-undefined-123456.example.com:5432/test_project"]}"
`;

exports[`branches > create branch and connect with psql and psql args 1`] = `
"branch:
  id: br-new-branch-123456
  name: test_branch
  created_at: 2021-01-01T00:00:00.000Z
endpoints:
  - id: ep-undefined-123456
    type: read_write
    created_at: 2021-01-01T00:00:00.000Z
    host: undefined.example.com
connection_uris:
  - connection_uri: postgres://ep-undefined-123456.example.com:5432/test_project
{"psql-cli-args":["postgres://ep-undefined-123456.example.com:5432/test_project","-c","SELECT 1"]}"
`;

exports[`branches > create by default with r/w endpoint 1`] = `
"branch:
  id: br-new-branch-123456
  name: test_branch
  created_at: 2021-01-01T00:00:00.000Z
endpoints:
  - id: ep-undefined-123456
    type: read_write
    created_at: 2021-01-01T00:00:00.000Z
    host: undefined.example.com
connection_uris:
  - connection_uri: postgres://ep-undefined-123456.example.com:5432/test_project
"
`;

exports[`branches > create with autoscaled CU 1`] = `
"id: br-new-branch-123456
name: test_branch_with_autoscaling
parent_id: br-main-branch-123456
created_at: 2021-01-01T00:00:00.000Z
"
`;

exports[`branches > create with fixed size CU 1`] = `
"id: br-new-branch-123456
name: test_branch_with_fixed_cu
parent_id: br-main-branch-123456
created_at: 2021-01-01T00:00:00.000Z
"
`;

exports[`branches > create with parent by lsn 1`] = `
"id: br-new-branch-123456
name: test_branch_with_parent_lsn
parent_id: br-main-branch-123456
created_at: 2021-01-01T00:00:00.000Z
"
`;

exports[`branches > create with parent by name 1`] = `
"id: br-new-branch-123456
name: test_branch_with_parent_name
parent_id: br-main-branch-123456
created_at: 2021-01-01T00:00:00.000Z
"
`;

exports[`branches > create with parent by timestamp 1`] = `
"id: br-new-branch-123456
name: test_branch_with_parent_timestamp
parent_id: br-main-branch-123456
created_at: 2021-01-01T00:00:00.000Z
"
`;

exports[`branches > create with readonly endpoint 1`] = `
"branch:
  id: br-new-branch-123456
  name: test_branch
  created_at: 2021-01-01T00:00:00.000Z
endpoints:
  - id: ep-undefined-123456
    type: read_only
    created_at: 2021-01-01T00:00:00.000Z
    host: undefined.example.com
connection_uris:
  - connection_uri: postgres://ep-undefined-123456.example.com:5432/test_project
"
`;

exports[`branches > create with suspend timeout 1`] = `
"id: br-new-branch-123456
name: test_branch_with_suspend_timeout
parent_id: br-main-branch-123456
created_at: 2021-01-01T00:00:00.000Z
suspend_timeout: 60
"
`;

exports[`branches > create without endpoint 1`] = `
"id: br-new-branch-123456
name: test_branch
created_at: 2021-01-01T00:00:00.000Z
"
`;

exports[`branches > delete by id 1`] = `
"id: br-sunny-branch-123456
name: sunny-branch
created_at: 2021-01-01T00:00:00.000Z
"
`;

exports[`branches > get by id 1`] = `
"name: test_branch
id: br-sunny-branch-123456
default: true
parent_id: br-parent-branch-123456
created_at: 2019-01-01T00:00:00Z
updated_at: 2019-01-01T00:00:00Z
"
`;

exports[`branches > get by id 2`] = `
"name: test-branch-cloudy
id: br-cloudy-branch-12345678
default: true
created_at: 2019-01-01T00:00:00Z
updated_at: 2019-01-01T00:00:00Z
"
`;

exports[`branches > get by name 1`] = `
"name: test_branch
id: br-sunny-branch-123456
default: true
parent_id: br-parent-branch-123456
created_at: 2019-01-01T00:00:00Z
updated_at: 2019-01-01T00:00:00Z
"
`;

exports[`branches > get by name with numeric name 1`] = `
"name: "123"
id: br-numbered-branch-123456
default: false
parent_id: br-parent-branch-123456
created_at: 2019-01-01T00:00:00Z
updated_at: 2019-01-01T00:00:00Z
"
`;

exports[`branches > list 1`] = `
"- id: br-main-branch-123456
  name: main
  default: true
  created_at: 2021-01-01T00:00:00.000Z
  updated_at: 2021-01-01T00:00:00.000Z
- id: br-sunny-branch-123456
  name: test_branch
  parent_id: br-parent-branch-123456
  created_at: 2021-01-01T00:00:00.000Z
  updated_at: 2021-01-01T00:00:00.000Z
- id: br-numbered-branch-123456
  name: "123"
  parent_id: br-parent-branch-123456
  created_at: 2021-01-01T00:00:00.000Z
  updated_at: 2021-01-01T00:00:00.000Z
- id: br-numbered-branch-123456
  name: test_branch_with_fixed_cu
  parent_id: br-parent-branch-123456
  created_at: 2021-01-01T00:00:00.000Z
  updated_at: 2021-01-01T00:00:00.000Z
- id: br-numbered-branch-123456
  name: test_branch_with_autoscaling
  parent_id: br-parent-branch-123456
  created_at: 2021-01-01T00:00:00.000Z
  updated_at: 2021-01-01T00:00:00.000Z
"
`;

exports[`branches > rename 1`] = `
"name: new_test_branch
created_at: 2021-01-01T00:00:00.000Z
"
`;

exports[`branches > reset branch to parent 1`] = `
"id: br-branch-123456
name: test-branch
source_branch_id: br-parent-branch-123456
project_id: pr-project-123456
created_at: 2020-01-01T00:00:00.000Z
updated_at: 2020-01-01T00:00:00.000Z
last_reset_at: 2020-01-01T00:00:00.000Z
"
`;

exports[`branches > restore branch to lsn 1`] = `
"id: br-self-tolsn-123456
name: self-tolsn
last_reset_at: 2021-01-01T00:00:00Z
"
`;

exports[`branches > restore to another branch head 1`] = `
"id: br-another-branch-123456
name: another-branch
last_reset_at: 2021-01-01T00:00:00Z
"
`;

exports[`branches > restore to parent branch timestamp by name 1`] = `
"id: br-parent-tots-123456
name: parent-tots
last_reset_at: 2021-01-01T00:00:00Z
"
`;

exports[`branches > restore with unexisted branch outputs error 1`] = `""`;

exports[`branches > set default by id 1`] = `
"name: test-branch-sunny
id: br-sunny-branch-123456
default: true
created_at: 2019-01-01T00:00:00Z
updated_at: 2019-01-01T00:00:00Z
"
`;

exports[`branches > set primary by id 1`] = `
"name: test-branch-sunny
id: br-sunny-branch-123456
default: true
created_at: 2019-01-01T00:00:00Z
updated_at: 2019-01-01T00:00:00Z
"
`;
